plugin/trino-exchange-filesystem/src/main/java/io/trino/plugin/exchange/filesystem/s3/S3FileSystemExchangeStorage.java

```diff
@@ -190,15 +224,26 @@ public ListenableFuture<Void> createEmptyFile(URI file)
     @Override
     public ListenableFuture<Void> deleteRecursively(URI dir)
     {
-        ImmutableList.Builder<String> keys = ImmutableList.builder();
-        return stats.getDeleteRecursively().record(transformFuture(Futures.transformAsync(
-                toListenableFuture((listObjectsRecursively(dir).subscribe(listObjectsV2Response ->
-                        listObjectsV2Response.contents().stream().map(S3Object::key).forEach(keys::add)))),
-                ignored -> {
-                    keys.add(keyFromUri(dir) + DIRECTORY_SUFFIX);
-                    return deleteObjects(getBucketName(dir), keys.build());
-                },
-                directExecutor())));
+        if (gcsClient.isPresent()) {
```

Comment:
nit: I would assign `compatibilityMode` to field and use it for dispatching (and only `verify` that `gcsClient.isPresent`. But I do not feel super strongly about that as so far `delete` impl is only GCS specific part.